module/extension/Director/src/bits/compare_priority.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+#include <algorithm>
+#include <tuple>
+#include <vector>
+
+#include "../Director.hpp"
+
+#include "extension/Behaviour.hpp"
+
+namespace module::extension {
+
+    using ::extension::behaviour::commands::DirectorTask;
+
+    bool Director::compare_priority(const std::shared_ptr<const DirectorTask>& incumbent,
+                                    const std::shared_ptr<const DirectorTask>& challenger) {
+
+        // Compare to yourself and you get that you shouldn't change tasks
+        if (incumbent->requester_id == challenger->requester_id) {
+            return false;
+        }
+
+        // Function to get the priorities of the ancestors of this task
+        auto get_ancestor_priorities = [this](const std::shared_ptr<const DirectorTask>& task) {
+            // Loop up through the providers until we reach a point where a task was emitted by a non provider
+            std::vector<std::tuple<uint64_t, int, bool>> ancestors;
+            for (auto t = task; providers.count(t->requester_id) != 0;) {
+                auto provider = providers[t->requester_id];
+                auto& group   = groups[provider->type];
+
+                if (group.active_task != nullptr) {
+                    ancestors.emplace_back(t->requester_id, t->priority, t->optional);
+                    t = group.active_task;
```

Comment:
I sort of get this function but not sure how it actually achieves what I think it's meant to do? I assume it goes up through the chain of provider task calls and stores all those tasks in `ancestors`. Is it also meant to include the input task `task`? 
And how does this line, `t=group.active_task` actually go up the chain? To me it looks like it just grabs the task that is active in this tasks group - surely that's the same task every time (except maybe not the first time)? 
My understanding of task groups is that you might have `statically_stable_walk` and `dynamically_stable_walk` and they're in a `walk` group together. Or maybe I'm getting confused with something else?
Definitely not getting something here, but I am curious about if the first task is an ancestor as well?

Reply:
I think there was a bug here, so I understand your confusion, I tried to explain it a little better in the comments.

Basically you get each tasks emitter, which will be different providers and you use them to go up the tree